### DISCUSSION TITLE
.ipynb ! command inside if block produces pylance error

### DISCUSSION THREAD

**Discussion by p-i-** [microsoft/pylance-release#4935] [Enhancement]
The following code in a .ipynb Python cell causes a PyLance error:

```
if 0:
    ! echo bar
```
The last letter has a red squiggly underline, with PyLance complaining: "Expected: Expression".

![Screenshot 2023-10-03 at 09 48 42](https://github.com/microsoft/vscode/assets/693495/3927e183-177d-46a3-947c-ddb6e29ad177)

Also note that it's very hard to actually click the buttons in that hover-box as it disappears when you move the mouse over to them (unless you're careful with your mouse movements).

**Comment by debonte** [comment]
Pylance does not currently treat `!`-prefixed system shell commands as expressions. For example, you also can't assign the results of such a command to a variable. https://ipython.org/ipython-doc/3/interactive/tutorial.html#system-shell-commands

Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by rubensa** [comment]
As @debonte said it is also not supported to assign the results of a shell command to a variable:

For example, for:
```python
dir = "/"

files = ! ls $dir
print(files)
```

I get the "Expected Expression" error message at the exclamation mark.

On the other hand, if the exclamation mark is the first character in a line, I don't get any error.

For example:

```python
dir = "/"

! ls $dir
```
